Navigation: Next button behaviour

Itxaka opened the pull request:
Uses the steps factory to track and set the current step finished to track if a step is finished or not and allow the next button to be enabled or disabled.

This is based on PR https://github.com/crowbar/crowbar-ui/pull/71 so it depends on that one to get merged.

Commit:
Create upgrade-nodes page template

Commit:
Update upgrade-nodes state in steps.factory and removed test copy from upgrade-nodes-page

Commit:
Implemented translation keys for the titles

Commit:
Moved refeshStepsList function inside the steps factory and updated next function.

santiph committed:
Remove trailing space

Commit:
Fix merge conflicts

Commit:
Modified nextStep function

Commit:
Modified text of comment line

Commit:
Merge branch 'features/upgrade7/steps-sidebar' of https://github.com/crowbar/crowbar-ui into features/upgrade7/steps-sidebar

Commit:
Add unit test to Crowbar-upgrade-steps directive

Commit:
Implemented unit test for upgrade.controller

Commit:
Merge branch 'features/upgrade7/steps-sidebar' of https://github.com/crowbar/crowbar-ui into features/upgrade7/steps-sidebar

Commit:
Use the steps factory to set and retrieve the current stet status

Itxaka commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(3) S49P20 #6 Layout - Navigation bar](https://trello.com/c/LNr710jO/104-3-s49p20-6-layout-navigation-bar)

skazi0 commented on assets/app/features/upgrade/upgrade.controller.js:
```diff
@@ -3,53 +3,31 @@
         // Get Steps list from provider
-        vm.steps.list = upgradeStepsFactory.getAll();
-        refeshStepsList();
+        vm.steps.list = upgradeStepsFactory.steps;
```
move it above to `vm.steps` initialization just like `stepFinished`.

Itxaka replied:
This actually comes from richa's PR, not this one. Stupid gituhb doent allow linked patches properly :/

skazi0 commented on assets/app/data/crowbar/steps.factory.js:
```diff
@@ -4,72 +4,107 @@
+
+        function refeshStepsList() {
+            var currentState = $state.current.name,
+                isCompletedStep = true;
```
It took me some time to understand the logic of `isCompletedStep`. Not sure if there's a better name for this but maybe it should at least be described in a comment that this flag indicates that given step was already completed which will be reflected in the UI.
This is especially confusing and a bit redundant when connected with `setCurrentStepCompleted()` above.

Itxaka replied:
This actually comes from richa's PR, not this one. Stupid gituhb doent allow linked patches properly :/

Also I think that var is not used anywhere for anything at all and was already in the code, I guess for preparation for tracking the step completition, but Im not sure I understand the reasoning of it at all. Should be removed in future patches as I dont think is useful.

santiph replied:
This var was used in the prototype to mark the completed steps in green. See https://github.com/crowbar/crowbar-ui/blob/master/assets/app/widgets/crowbar-upgrade-steps/crowbar-upgrade-steps.directive.jade#L2

It's a miss from the cleanup card and will be removed on Richa's changes

skazi0 commented on assets/app/features/upgrade/upgrade.controller.js:
```diff
@@ -59,22 +37,16 @@
-            vm.steps.activeStep.enabled = true;
 
-            $state.go(vm.steps.activeStep.state);
+            $state.go(vm.steps.list[upgradeStepsFactory.activeStep.id + 1].state);
```
The whole `nextStep()` and `isLastStep()` belong in `upgradeStepsFactory`. The factory already uses `$state` and owns the steps list. They can be linked here just like the rest of `vm.steps`.

Itxaka replied:
This actually comes from richa's PR, not this one. Stupid gituhb doent allow linked patches properly :/

skazi0 commented on assets/app/features/upgrade/upgrade.controller.js:
```diff
@@ -3,53 +3,31 @@
             nextStep: nextStep,
-            isLastStep: isLastStep
+            isLastStep: isLastStep,
+            stepFinished: upgradeStepsFactory.stepFinished
```
Not sure what our convention is but we should try to keep the functions returning booleans named consistently. Here we have `isLastStep()` and `stepFinished()`.

Itxaka commented:
I kept breaking my git history while trying to rebase this properly, so I just extracted the commit, updated it and pushed a new PR: https://github.com/crowbar/crowbar-ui/pull/81
